File: packages/javascript/lib/math/numUnique.ts
// Last value handed out, so repeated calls can never collide.
let lastId = 0;

export function numUnique(): number {
	// Milliseconds (13 digits) * 1000 leaves room for a per-millisecond sequence while
	// staying inside Number.MAX_SAFE_INTEGER. The previous 18-digit value exceeded it, so
	// digits were rounded away and different draws collapsed onto the same number.
	const id = new Date().valueOf() * 1000;

	// Always move forward: within the same millisecond, and even if the system clock
	// steps backwards.
	lastId = id > lastId ? id : lastId + 1;

	return lastId;
}

File: packages/python/qsu/math/numUnique.py
import time

# Last value handed out, so repeated calls can never collide.
_lastId = 0


def numUnique() -> int:
	"""Return a unique number based on the current timestamp.

	Milliseconds * 1000 leaves room for a per-millisecond sequence, so repeated calls
	within a process always return a new, strictly increasing value. Uniqueness is only
	guaranteed within one process, and the value is sequential and therefore predictable,
	so it must not be used for anything security related.
	"""
	global _lastId

	id = int(time.time() * 1000) * 1000

	# Always move forward: within the same millisecond, and even if the clock steps back.
	_lastId = id if id > _lastId else _lastId + 1

	return _lastId
